File: mcp-server/src/simulator.ts
#!/usr/bin/env node
/**
 * Amiga Simulator - Pretends to be an Amiga app connected via serial TCP.
 * Runs as a TCP server on port 1234 (or AMIGA_SERIAL_PORT).
 * Sends realistic protocol messages and responds to host commands.
 *
 * Usage: npx tsx src/simulator.ts
 */

import { createServer, Socket } from "net";

const PORT = parseInt(process.env.AMIGA_SERIAL_PORT ?? "1234", 10);

// Simulated state
let tick = 0;
let ball_x = 160;
let ball_y = 100;
let ball_dx = 3;
let ball_dy = 2;
let frame_count = 0;
let freeChip = 512000;
let freeFast = 1048576;
let running = true;

// Variable registry for GETVAR/SETVAR
const vars: Record<string, { type: string; get: () => string; set?: (v: string) => void }> = {
  ball_x:      { type: "i32", get: () => String(ball_x),      set: (v) => { ball_x = parseInt(v); } },
  ball_y:      { type: "i32", get: () => String(ball_y),      set: (v) => { ball_y = parseInt(v); } },
  ball_dx:     { type: "i32", get: () => String(ball_dx),     set: (v) => { ball_dx = parseInt(v); } },
  ball_dy:     { type: "i32", get: () => String(ball_dy),     set: (v) => { ball_dy = parseInt(v); } },
  frame_count: { type: "u32", get: () => String(frame_count) },
  free_chip:   { type: "u32", get: () => String(freeChip) },
  free_fast:   { type: "u32", get: () => String(freeFast) },
};

// Simulated memory (just some patterns)
function getMemory(addr: number, size: number): string {
  let hex = "";
  for (let i = 0; i < size; i++) {
    const byte = (addr + i) & 0xFF;
    hex += byte.toString(16).padStart(2, "0").toUpperCase();
  }
  return hex;
}

function sendLine(socket: Socket, line: string) {
  try {
    socket.write(line + "\n");
  } catch { /* ignore write errors */ }
}

function handleCommand(socket: Socket, line: string) {
  const parts = line.split("|");
  const cmd = parts[0];

  switch (cmd) {
    case "PING":
      sendLine(socket, `HB|${tick}|${freeChip}|${freeFast}`);
      break;

    case "GETVAR":
      if (parts.length >= 2) {
        const name = parts[1];
        const v = vars[name];
        if (v) {
          sendLine(socket, `VAR|${name}|${v.type}|${v.get()}`);
        } else {
          sendLine(socket, `VAR|${name}|err|not_found`);
        }
      }
      break;

    case "SETVAR":
      if (parts.length >= 3) {
        const name = parts[1];
        const value = parts[2];
        const v = vars[name];
        if (v && v.set) {
          v.set(value);
          sendLine(socket, `VAR|${name}|${v.type}|${v.get()}`);
          sendLine(socket, `LOG|I|${tick}|Host set ${name} = ${value}`);
        }
      }
      break;

    case "INSPECT":
      if (parts.length >= 3) {
        const addr = parseInt(parts[1], 16);
        let size = parseInt(parts[2]);
        if (size > 4096) size = 4096;
        // Send in 256-byte chunks
        let offset = 0;
        while (offset < size) {
          const chunk = Math.min(256, size - offset);
          const hexData = getMemory(addr + offset, chunk);
          sendLine(socket, `MEM|${(addr + offset).toString(16).padStart(8, "0")}|${chunk}|${hexData}`);
          offset += chunk;
        }
      }
      break;

    case "EXEC":
      if (parts.length >= 3) {
        const id = parts[1];
        const expression = parts.slice(2).join("|");
        if (expression === "reset") {
          ball_x = 160;
          ball_y = 100;
          ball_dx = 3;
          ball_dy = 2;
          sendLine(socket, `CMD|${id}|ok|Ball position reset`);
          sendLine(socket, `LOG|I|${tick}|Ball position reset by host`);
        } else if (expression === "status") {
          sendLine(socket, `CMD|${id}|ok|ball(${ball_x},${ball_y}) vel(${ball_dx},${ball_dy}) frame=${frame_count}`);
        } else {
          sendLine(socket, `CMD|${id}|ok|Executed: ${expression}`);
        }
      }
      break;
  }
}

const tcpServer = createServer((socket) => {
  console.log(`MCP host connected from ${socket.remoteAddress}:${socket.remotePort}`);

  let lineBuf = "";

  // Send startup messages
  sendLine(socket, `LOG|I|${tick++}|Debug session started`);
  sendLine(socket, `HB|${tick}|${freeChip}|${freeFast}`);
  sendLine(socket, `LOG|I|${tick++}|Bouncing Ball demo starting`);
  sendLine(socket, `LOG|I|${tick++}|Window opened: inner 312x178`);

  // Simulation loop - runs the "bouncing ball" logic
  const simInterval = setInterval(() => {
    if (!running) return;

    // Update ball position
    ball_x += ball_dx;
    ball_y += ball_dy;

    // Bounce
    if (ball_x <= 5 || ball_x >= 307) {
      ball_dx = -ball_dx;
      ball_x += ball_dx;
      sendLine(socket, `LOG|D|${tick}|Bounce X at ${ball_x}`);
    }
    if (ball_y <= 5 || ball_y >= 173) {
      ball_dy = -ball_dy;
      ball_y += ball_dy;
      sendLine(socket, `LOG|D|${tick}|Bounce Y at ${ball_y}`);
    }

    frame_count++;
    tick++;

    // Vary free memory slightly to make it realistic
    freeChip += Math.floor(Math.random() * 200) - 100;
    freeFast += Math.floor(Math.random() * 400) - 200;

    // Heartbeat every 60 frames
    if (frame_count % 60 === 0) {
      sendLine(socket, `HB|${tick}|${freeChip}|${freeFast}`);
    }

    // Periodic info logs
    if (frame_count % 120 === 0) {
      sendLine(socket, `LOG|I|${tick}|frame=${frame_count} ball(${ball_x},${ball_y}) vel(${ball_dx},${ball_dy})`);
    }

    // Occasional warnings (memory getting low simulation)
    if (frame_count % 500 === 0 && freeChip < 480000) {
      sendLine(socket, `LOG|W|${tick}|Chip memory getting low: ${freeChip} bytes`);
    }
  }, 50); // ~20fps like the real app

  // Handle incoming commands from host
  socket.on("data", (data) => {
    lineBuf += data.toString();
    const lines = lineBuf.split("\n");
    lineBuf = lines.pop() ?? "";

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      console.log(`  <- ${trimmed}`);
      handleCommand(socket, trimmed);
    }
  });

  socket.on("close", () => {
    console.log("MCP host disconnected");
    clearInterval(simInterval);
  });

  socket.on("error", () => {
    clearInterval(simInterval);
  });
});

tcpServer.listen(PORT, () => {
  console.log(`\n=== Amiga Simulator ===`);
  console.log(`Listening on TCP port ${PORT}`);
  console.log(`Simulating: Bouncing Ball demo`);
  console.log(`Variables: ${Object.keys(vars).join(", ")}`);
  console.log(`Commands: reset, status`);
  console.log(`\nWaiting for MCP server to connect...\n`);
});

tcpServer.on("error", (err) => {
  console.error(`Server error: ${err.message}`);
  process.exit(1);
});

File: amiga-devbench/amiga_devbench/simulator.py
"""Amiga simulator - TCP server that pretends to be an Amiga running a bouncing ball app."""

from __future__ import annotations

import asyncio
import logging
import math
import random

logger = logging.getLogger(__name__)


class AmigaSimulator:
    """TCP server on a given port that fakes an Amiga with a bouncing ball demo."""

    def __init__(self, port: int = 1234) -> None:
        self._port = port
        self._server: asyncio.Server | None = None

        # Simulated state
        self._tick = 0
        self._ball_x = 160
        self._ball_y = 100
        self._ball_dx = 3
        self._ball_dy = 2
        self._frame_count = 0
        self._free_chip = 512000
        self._free_fast = 1048576
        self._running = True

        # Variable registry
        self._vars: dict[str, dict] = {
            "ball_x": {"type": "i32", "get": lambda: str(self._ball_x),
                       "set": lambda v: setattr(self, "_ball_x", int(v))},
            "ball_y": {"type": "i32", "get": lambda: str(self._ball_y),
                       "set": lambda v: setattr(self, "_ball_y", int(v))},
            "ball_dx": {"type": "i32", "get": lambda: str(self._ball_dx),
                        "set": lambda v: setattr(self, "_ball_dx", int(v))},
            "ball_dy": {"type": "i32", "get": lambda: str(self._ball_dy),
                        "set": lambda v: setattr(self, "_ball_dy", int(v))},
            "frame_count": {"type": "u32", "get": lambda: str(self._frame_count)},
            "free_chip": {"type": "u32", "get": lambda: str(self._free_chip)},
            "free_fast": {"type": "u32", "get": lambda: str(self._free_fast)},
        }

    def _get_memory(self, addr: int, size: int) -> str:
        """Generate fake memory contents."""
        return "".join(f"{(addr + i) & 0xFF:02X}" for i in range(size))

    async def start(self) -> None:
        self._server = await asyncio.start_server(
            self._handle_client, "127.0.0.1", self._port,
        )
        logger.info("Amiga Simulator listening on TCP port %d", self._port)
        logger.info("Simulating: Bouncing Ball demo")
        logger.info("Variables: %s", ", ".join(self._vars.keys()))

    async def stop(self) -> None:
        self._running = False
        if self._server:
            self._server.close()
            await self._server.wait_closed()
            self._server = None

    async def _handle_client(
        self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter,
    ) -> None:
        addr = writer.get_extra_info("peername")
        logger.info("MCP host connected from %s", addr)

        def send_line(line: str) -> None:
            try:
                writer.write((line + "\n").encode("latin-1"))
            except Exception:
                pass

        # Send startup messages
        send_line(f"LOG|I|{self._tick}|Debug session started")
        self._tick += 1
        send_line(f"HB|{self._tick}|{self._free_chip}|{self._free_fast}")
        send_line(f"LOG|I|{self._tick}|Bouncing Ball demo starting")
        self._tick += 1
        send_line(f"LOG|I|{self._tick}|Window opened: inner 312x178")
        self._tick += 1

        sim_task = asyncio.ensure_future(self._simulation_loop(send_line))
        line_buf = ""

        try:
            while True:
                data = await reader.read(4096)
                if not data:
                    break
                line_buf += data.decode("latin-1", errors="replace")
                lines = line_buf.split("\n")
                line_buf = lines.pop()

                for line in lines:
                    trimmed = line.strip()
                    if not trimmed:
                        continue
                    logger.debug("  <- %s", trimmed)
                    self._handle_command(send_line, trimmed)
                    await writer.drain()
        except (asyncio.CancelledError, ConnectionError):
            pass
        finally:
            sim_task.cancel()
            try:
                await sim_task
            except asyncio.CancelledError:
                pass
            writer.close()
            try:
                await writer.wait_closed()
            except Exception:
                pass
            logger.info("MCP host disconnected")

    async def _simulation_loop(self, send_line) -> None:
        """Run the bouncing ball simulation at ~20fps."""
        try:
            while self._running:
                await asyncio.sleep(0.05)  # ~20fps

                # Update ball
                self._ball_x += self._ball_dx
                self._ball_y += self._ball_dy

                if self._ball_x <= 5 or self._ball_x >= 307:
                    self._ball_dx = -self._ball_dx
                    self._ball_x += self._ball_dx
                    send_line(f"LOG|D|{self._tick}|Bounce X at {self._ball_x}")

                if self._ball_y <= 5 or self._ball_y >= 173:
                    self._ball_dy = -self._ball_dy
                    self._ball_y += self._ball_dy
                    send_line(f"LOG|D|{self._tick}|Bounce Y at {self._ball_y}")

                self._frame_count += 1
                self._tick += 1

                # Vary memory slightly
                self._free_chip += random.randint(-100, 100)
                self._free_fast += random.randint(-200, 200)

                # Heartbeat every 60 frames
                if self._frame_count % 60 == 0:
                    send_line(f"HB|{self._tick}|{self._free_chip}|{self._free_fast}")

                # Periodic info logs
                if self._frame_count % 120 == 0:
                    send_line(
                        f"LOG|I|{self._tick}|frame={self._frame_count} "
                        f"ball({self._ball_x},{self._ball_y}) "
                        f"vel({self._ball_dx},{self._ball_dy})"
                    )

                # Occasional warnings
                if self._frame_count % 500 == 0 and self._free_chip < 480000:
                    send_line(
                        f"LOG|W|{self._tick}|Chip memory getting low: {self._free_chip} bytes"
                    )
        except asyncio.CancelledError:
            pass

    def _handle_command(self, send_line, line: str) -> None:
        parts = line.split("|")
        cmd = parts[0]

        if cmd == "PING":
            send_line(f"HB|{self._tick}|{self._free_chip}|{self._free_fast}")

        elif cmd == "GETVAR" and len(parts) >= 2:
            name = parts[1]
            v = self._vars.get(name)
            if v:
                send_line(f"VAR|{name}|{v['type']}|{v['get']()}")
            else:
                send_line(f"VAR|{name}|err|not_found")

        elif cmd == "SETVAR" and len(parts) >= 3:
            name = parts[1]
            value = parts[2]
            v = self._vars.get(name)
            if v and "set" in v:
                v["set"](value)
                send_line(f"VAR|{name}|{v['type']}|{v['get']()}")
                send_line(f"LOG|I|{self._tick}|Host set {name} = {value}")

        elif cmd == "INSPECT" and len(parts) >= 3:
            addr = int(parts[1], 16)
            size = min(int(parts[2]), 4096)
            offset = 0
            while offset < size:
                chunk = min(256, size - offset)
                hex_data = self._get_memory(addr + offset, chunk)
                send_line(f"MEM|{(addr + offset):08x}|{chunk}|{hex_data}")
                offset += chunk

        elif cmd == "EXEC" and len(parts) >= 3:
            cmd_id = parts[1]
            expression = "|".join(parts[2:])
            if expression == "reset":
                self._ball_x = 160
                self._ball_y = 100
                self._ball_dx = 3
                self._ball_dy = 2
                send_line(f"CMD|{cmd_id}|ok|Ball position reset")
                send_line(f"LOG|I|{self._tick}|Ball position reset by host")
            elif expression == "status":
                send_line(
                    f"CMD|{cmd_id}|ok|ball({self._ball_x},{self._ball_y}) "
                    f"vel({self._ball_dx},{self._ball_dy}) frame={self._frame_count}"
                )
            else:
                send_line(f"CMD|{cmd_id}|ok|Executed: {expression}")

        elif cmd == "LISTCLIENTS":
            send_line("CLIENTS|1|bouncing_ball")

        elif cmd == "LISTTASKS":
            import json
            tasks = [
                {"name": "bouncing_ball", "priority": 0, "state": "running", "stackSize": 4096},
                {"name": "input.device", "priority": 20, "state": "waiting", "stackSize": 4096},
                {"name": "trackdisk.device", "priority": 5, "state": "waiting", "stackSize": 4096},
            ]
            send_line(f"TASKS|{len(tasks)}|{json.dumps(tasks)}")

        elif cmd == "LISTLIBS":
            import json
            libs = [
                {"name": "exec.library", "version": 40, "revision": 10},
                {"name": "dos.library", "version": 40, "revision": 3},
                {"name": "intuition.library", "version": 40, "revision": 5},
                {"name": "graphics.library", "version": 40, "revision": 2},
            ]
            send_line(f"LIBS|{len(libs)}|{json.dumps(libs)}")

        elif cmd == "LISTDIR" and len(parts) >= 2:
            import json
            path = parts[1]
            entries = [
                {"name": "c", "type": "dir", "size": 0, "date": "2024-01-15", "prot": "----rwed", "path": f"{path}c/"},
                {"name": "libs", "type": "dir", "size": 0, "date": "2024-01-15", "prot": "----rwed", "path": f"{path}libs/"},
                {"name": "startup-sequence", "type": "file", "size": 512, "date": "2024-01-15", "prot": "----rwed"},
            ]
            send_line(f"DIR|{path}|{len(entries)}|{json.dumps(entries)}")

        elif cmd == "READFILE" and len(parts) >= 4:
            path = parts[1]
            offset = int(parts[2])
            size = min(int(parts[3]), 4096)
            hex_data = self._get_memory(offset, size)
            send_line(f"FILE|{path}|{size}|{offset}|{hex_data}")

        elif cmd == "LAUNCH" and len(parts) >= 3:
            cmd_id = parts[1]
            command = "|".join(parts[2:])
            send_line(f"PROC|{cmd_id}|started|Launched: {command}")
            send_line(f"LOG|I|{self._tick}|Launched process: {command}")

        elif cmd == "DOSCOMMAND" and len(parts) >= 3:
            cmd_id = parts[1]
            command = "|".join(parts[2:])
            send_line(f"CMD|{cmd_id}|ok|DOS: {command}")
